gnc/sub8_perception/src/sub8_pcl/cv_utils.cc

```diff
@@ -1,12 +1,270 @@
+	// 	}
+	// }
+	// else std::cout << "same as 'Maxima'";
+	// std::cout << std::endl;
+	return threshed_local_maxima;
+}
+
+
+std::vector<cv::Point> find_local_minima(const cv::MatND &histogram, float thresh_multiplier){
+	std::vector<cv::Point> local_minima, threshed_local_minima;
+	float global_minimum = 0;
+
+	// Locate local minima and find global minimum
+	for(size_t idx = 1; idx < histogram.total() - 1; idx++){
+		float current_value = histogram.at<float>(idx);
+	    if((histogram.at<float>(idx-1) >= current_value) && (histogram.at<float>(idx+1) > current_value)){
+	    	local_minima.push_back(cv::Point(idx, current_value));
+	    	if(global_minimum > current_value) global_minimum = current_value;
+	    }
+	}
+	// std::cout << std::endl << "Minima: ";
+	BOOST_FOREACH(cv::Point pt, local_minima){
+		if(pt.y < global_minimum * thresh_multiplier) threshed_local_minima.push_back(pt);
+		// std::cout << boost::format("[%1%, %2%] ") % pt.x % pt.y;
+	}
+	// std::cout << std::endl << boost::format("thresh: < global_minimum(%1%) * thresh_multiplier(%2%) = %3%") 
+	// 							% global_minimum % thresh_multiplier % (global_minimum * thresh_multiplier);
+	// std::cout << std::endl << "Threshed Minima (x): ";
+	// if (threshed_local_minima.size() != local_minima.size()){
+	// 	BOOST_FOREACH(cv::Point pt, threshed_local_minima){
+	// 		std::cout << boost::format(" %1% ") % pt.x;
+	// 	}
+	// }
+	// else std::cout << "same as 'Minima'";
+	// std::cout << std::endl;
+	return threshed_local_minima;
+}
+
+
+unsigned int select_hist_mode(std::vector<cv::Point> &histogram_modes, int target){
+  std::vector<int> distances;
+  BOOST_FOREACH(cv::Point mode, histogram_modes){
+    distances.push_back(mode.x - target);
+  }
+  int min_idx = 0;
+  for(int i = 0; i < distances.size(); i++){
+    if(std::abs(distances[i]) <= std::abs(distances[min_idx])) min_idx = i;
+  }
+  return histogram_modes[min_idx].x;
+}
+
+
+void statistical_image_segmentation(const cv::Mat &src, cv::Mat &dest, const int hist_size,
+        							const float** ranges, const int target, std::string image_name, 
+        							const float sigma, const float low_thresh_gain, const float high_thresh_gain)
+{
+	// Calculate histogram
+	cv::MatND hist, hist_smooth, hist_derivative;
+	cv::calcHist( &src, 1, 0, cv::Mat(), hist, 1, &hist_size, ranges, true, false );
+
+	// Smooth histogram
+	const int kernel_size = 11;
+	hist_smooth = sub::smooth_histogram(hist, kernel_size, sigma);
+
+	// Calculate histogram derivative (central finite difference)
+	hist_derivative = hist_smooth.clone();
+	hist_derivative.at<float>(0) = 0;
+	hist_derivative.at<float>(hist_size - 1) = 0;
+	for (int i = 1; i < hist_size - 1; ++i)
+	{
+		hist_derivative.at<float>(i) = (hist_smooth.at<float>(i + 1) - hist_smooth.at<float>(i - 1)) / 2.0;
+	}
+	hist_derivative = sub::smooth_histogram(hist_derivative, kernel_size, sigma);
+
+	// Find target mode
+	// std::cout << boost::format("Target: %1%") % target;
```

Comment:
The only reservation I have is that these commented debug statements will likely prove very useful for quick tuning at the competition. What do you say?